src/lib/peers/peer-connection/PeerConnection.ts

```diff
@@ -0,0 +1,239 @@
+  private credentials: grpc.ChannelCredentials;
+
+  constructor(publicKey: string, keyManager: KeyManager, peerManager: PeerManager) {
+    this.publicKey = publicKey;
+    this.keyManager = keyManager;
+    this.peerManager = peerManager;
+
+    const pkiInfo = keyManager.PKIInfo;
+    if (pkiInfo.caCert && pkiInfo.cert && pkiInfo.key) {
+      this.credentials = grpc.credentials.createSsl(pkiInfo.caCert, pkiInfo.key, pkiInfo.cert);
+    } else {
+      this.credentials = grpc.credentials.createInsecure();
+    }
+  }
+
+  private async connectDirectly(): Promise<PeerClient> {
+    // try to create a direct connection
+    if (this.getPeerInfo().connectedAddress) {
+      // direct connection attempt
+      const address = this.getPeerInfo().connectedAddress!;
+      const peerClient = new PeerClient(address.toString(), this.credentials);
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().connectedAddress) {
+      throw Error('peer does not have a connected address');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  private async connectHolePunch(): Promise<PeerClient> {
+    // try to hole punch to peer via relay peer
+    if (!this.connected && this.getPeerInfo().relayPublicKey) {
+      // connect to relay and ask it to create a relay
+      console.log('requesting udp hole punch connection');
+      const connectedAddress = await this.peerManager.turnClient.requestHolePunchConnection(
+        this.getPeerInfo().relayPublicKey!,
+        this.getPeerInfo().publicKey,
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  private async connectRelay(): Promise<PeerClient> {
+    // try to relay to peer via relay peer
+    if (!this.connected && this.getPeerInfo().relayPublicKey) {
+      // turn relay
+      // connect to relay and ask it to create a relay
+      const connectedAddress = await this.peerManager.turnClient.requestPeerConnection(
+        this.getPeerInfo().publicKey,
+        this.getPeerInfo().relayPublicKey!,
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  async connectFirstChannel() {
+    return await new Promise<PeerClient>((resolve, reject) => {
+      const promiseList = [this.connectDirectly(), this.connectHolePunch(), this.connectRelay()];
+
+      const errorList: Error[] = [];
+      for (const promise of promiseList) {
+        promise
```

Comment:
Even though these are promises, can you continue to use async await here? You can `await` the promise right?

Reply:
well part of this script is a promise utility: resolving on the first promise and rejecting if all reject. This is an attempt to try and connect to the first successful address and only reject if none work. This could probably be brought out into a utility function. As for async await, I can see a way to do that now with trycatch blocks so I will try that now to see if it achieves the same thing.

Reply:
I remember now, async/await doesn't work here. The point is you don't want to wait for any particular promise, you want them to race as in `Promise.race`. The only issue with that convenience function is that it rejects when any promise rejects as well, here we want a resolve race but not a reject race.

The `bluebird` promise library has this functionality in the form of `Promise.any(..)` but I decided to just implement this ourselves to save a dependency.

Reply:
I've added this to the utils so now it looks like this:
```
async connectFirstChannel() {
  return await firstPromiseFulfilled([
    this.connectDirectly(),
    this.connectHolePunch(),
    this.connectRelay()
  ])
}
```